src/EnergyPlus/General.cc

```diff
@@ -288,6 +288,26 @@ void SolveRoot(const EnergyPlusData &state,
     XRes = XTemp;
 }
 
+void MovingAvg(EPVector<Real64> &DataIn, int const NumItemsInAvg)
+{
+    if (NumItemsInAvg <= 1) return; // no need to average/smooth
+
+    EPVector<Real64> TempData;
+    TempData.allocate(2 * DataIn.size()); // a scratch array twice the size, bottom end duplicate of top end
+
+    for (std::size_t i = 1; i <= DataIn.size(); ++i) {
+        TempData(i) = TempData(DataIn.size() + i) = DataIn(i); // initialize both bottom and top end
+        DataIn(i) = 0.0;
+    }
+
+    for (std::size_t i = 1; i <= DataIn.size(); ++i) {
+        for (int j = 1; j <= NumItemsInAvg; ++j) {
+            DataIn(i) += TempData(DataIn.size() - NumItemsInAvg + i + j); // sum top end including NumItemsInAvg history terms
+        }
+        DataIn(i) /= NumItemsInAvg; // average to smooth over NumItemsInAvg window
+    }
+}
```

Comment:
Question 1:
How often is this function being called?  Calling allocate a lot doesn't make me feel great.  Is it possible to create a scratch vector on state and when you get inside here, just overwrite the indices of interest and ignore the rest of the vector?  Then we wouldn't have to continually call allocate each time we get in here?  It may not be worth changing to that, but thought it was worth bringing it up.

Question 2:
Can we just make one MovingAvg function that accepts a gsl::span argument?  There has to be a nice way to generalize the functionality so that it can operate on either an Array or a vector regardless of the base index.  It's  just a series of variables from M to N.  [Here's](https://godbolt.org/) an example of using span for both array and vector which does not reference the zero index.  I get that there are two issues: whether an Objexx::Array1D be passed as a span, and how complex would it be to handle the array index values generically.  So this could easily just be pinned for investigation later, or we can talk about it and I can try to hack around and see what I can get done in a few minutes.

Reply:
There is a faster way to do this algorithmically. One that uses a single loop, not a nested loop. And it's possible that this faster way also allows you to eliminate the memory allocation. I'll leave this as an exercise to the reader.

Reply:
Ok, so I had to do this to convince myself that it works. You still need a temporary array but that array is of size `NumItemsInAvg` and so maybe that can be a `std::array` and `NumItemsInAvg` can be a function template variable.

Reply:
I thought I had a working solution that passes the `General_MovingAvg` unit test, but then it failed for full simulations. Not sure what's inadequate about the unit test, but I'm giving up. I'm willing to copy your work if you're wiling to share it. Meet me in the cloak room so the teacher doesn't find out.

Reply:
The other use of `MovingAvg` is [here](https://github.com/NREL/EnergyPlus/blob/f1db7f411c004a7e0122a0cf77bc9c03d3a59183/src/EnergyPlus/OutputReportTabular.cc#L15509-L15521) where an `Array1D` is created from a slice of an `Array3D`:
```
        AvgData = state.dataOutRptTab->peopleInstantSeq(desDaySelected, _, zoneIndex);
        General::MovingAvg(AvgData, state.dataSize->NumTimeStepsInAvg);
```
This pattern repeats 27 times. This is ripe for a refactor, but not here.
@Myoldmopar I can leave this as-is in this branch, or revert some of the sizing `EPVectors` to `Array1D` and delete the duplicate `MovingAvg` function. Let me know your preference.

Reply:
Ok, so I tried a few files and got either no diffs or diffs in the 14th:

```
657063,657065c657063,657065
< 1469,13.352493415855038
< 1470,0.0008620320215328262
< 1471,-16.91673335013139
---
> 1469,13.352493415855047
> 1470,0.0008620320215328258
> 1471,-16.916733350131434
```

Here is my implementation:

```c++
void MovingAvg(Array1D<Real64> &DataIn, int const NumItemsInAvg)
{
    if (NumItemsInAvg <= 1) return; // no need to average/smooth                                                                                              

    Array1D<Real64> movingWindow(NumItemsInAvg);     // holds the last NumItemsInAvg values                                                                                                     
    Real64 movingSum = 0.0; // holds the sum of movingWindow                                                                                                               

    // Initialize movingWindow with last NumItemsInAvg - 1 elements of DataIn                                                                                                                               
    int iWindow;
    for (iWindow = 1; iWindow <= NumItemsInAvg - 1; ++iWindow) {
        movingWindow(iWindow) = DataIn(DataIn.size() - (NumItemsInAvg - 1) + iWindow);
        movingSum += movingWindow(iWindow);
    }

    // For every new element, slide movingWindow and movingSum forward one.                                                                                   
    for (int i = 1; i <= (int)DataIn.size(); ++i) {
        movingWindow(iWindow) = DataIn(i);
        movingSum += movingWindow(iWindow);
        DataIn(i) = movingSum / NumItemsInAvg;
        // movingWindow is indexed in circular fashion                                                                                                        
        ++iWindow; if (iWindow > NumItemsInAvg) iWindow = 1;
        // Subtract value of oldest element in movingWindow from movingSum                                                
        movingSum -= movingWindow(iWindow);     
    }
}
```

Reply:
Moot now, but for the record, unit test `AirLoopHVACDOAS_TestFanHeatAddeToCoolingCoilSize` and regression test 5ZoneCoolBeam (among others) failed while unit test `General_MovingAvg` passed.  Here was my attempt:
```
void MovingAvg(Array1D<Real64> &dataIn, int const numItemsInAvg)
{
    if (numItemsInAvg <= 1) return; // no need to average/smooth

    std::vector<Real64> tempData(2 * numItemsInAvg - 1); // temp data
    int size = dataIn.size();
    int tailStart = size - (numItemsInAvg - 1); // Need to save the last (numItemsInAvg-1) terms

    for (int i = 0; i < numItemsInAvg - 1; ++i) {
        tempData[i] = dataIn[tailStart + i]; // save last items to use later
    }
    for (int i = 0; i < numItemsInAvg; ++i) {
        tempData[i + numItemsInAvg - 1] = dataIn[i]; // save first items to use later
    }

    // Start at end and compute all of the averages except the first few that need the end terms
    for (int i = size; i > numItemsInAvg; --i) {
        dataIn[i - 1] = std::accumulate(dataIn.begin() + (i - numItemsInAvg), dataIn.begin() + i, 0) / (Real64)numItemsInAvg;
    }

    // Start at the begnning and compute the averages for the first few terms using the saved tempData
    for (int i = 0; i <= numItemsInAvg - 1; ++i) {
        dataIn[i] = std::accumulate(tempData.begin() + i, tempData.begin() + (i + numItemsInAvg), 0) / (Real64)numItemsInAvg;
    }
}
```

Reply:
And now there are 2 new suggestions for moving average. I wonder which is faster?

Reply:
Mine I think. `std::accumulate` is itself a loop and I don't use nested loops.

Reply:
That unit test passes for me. I get some diffs in `5ZoneCoolBeam` but only in the 14th and 15th decimal places again. Whenever you change order and/or grouping of floating-point operations, you're going to get different roundings here and there. No escaping this in a fixed-precision representation.

Reply:
Diffs in 14th/15th place is fine. For whatever reason, mine was getting large diffs, e.g. the airloop flow was zero in `AirLoopHVACDOAS_TestFanHeatAddeToCoolingCoilSize`. Doesn't really matter. I'll use @amirroth 's and move on.
And I'm going to revert some `EPVectors` in DataSizing so that only one flavor of `MovingAvg` is needed.

Reply:
Also, mine is just so cute! Look at it! The `movingWindow` array is a perfect circle! The British Museum uses it to calibrate their instruments!

Reply:
That new algorithm is making my head hurt. This first loop initializes the temp array to the top end of the data array. And then sums those data:

    for (iWindow = 1; iWindow <= NumItemsInAvg - 1; ++iWindow) {
        movingWindow(iWindow) = DataIn(DataIn.size() - (NumItemsInAvg - 1) + iWindow);
        movingSum += movingWindow(iWindow);
    }

And then in the next loop, a new data point is introduced, and then added to the sum of the data in the initial averaging window (NumItemsInAvg). Let's say NumItemsInAvg = 4. The loop above sums 4 values. And now here, the first element in the array is added to the sum of the top 4 elements. Isn't that 5 elements in a window of 4 data points? I realize that diffs are 1E-14, but this just looks strange. I am probably missing a key point here.

    // For every new element, slide movingWindow and movingSum forward one.                                                                                   
    for (int i = 1; i <= (int)DataIn.size(); ++i) {
        movingWindow(iWindow) = DataIn(i);
        movingSum += movingWindow(iWindow);
        DataIn(i) = movingSum / NumItemsInAvg;

Reply:
The initialization loop only sums `NumItemsInAvg - 1` values. In your example it would only sum `3`.

Reply:
Ah, there it is, `for (iWindow = 1; iWindow <= NumItemsInAvg - 1;`

Reply:
Does it complicate things or simplify things to note that Array1D objects can be accessed as a zero-based vector using the square brackets?  I wonder, if this function was written as zero based using square brackets, could the function just be a template to take in either a vector or an Array1D?  This is just conversation, we don't need to dwell on it any longer.

Reply:
`Array1D` or `EPVector`? If you wanted to handle both `vector` or `EPVector` you  could use a template or `span` and square brackets.

Reply:
> `Array1D` or `EPVector`? If you wanted to handle both `vector` or `EPVector` you could use a template or `span` and square brackets.

Probably, but I don't speak span or template very well. The bigger issue is getting rid of the tiny values that are causing big diffs. I may just revert all of this out of here and put it in a new branch so the space stuff can move forward.